oauth2_http/java/com/google/auth/oauth2/AwsCredentials.java

```diff
@@ -0,0 +1,348 @@
+    token.put(
+        "url",
+        ((AwsCredentialSource) credentialSource)
+            .regionalCredentialVerificationUrl.replace("{region}", signature.getRegion()));
+    return URLEncoder.encode(token.toString(), "UTF-8");
+  }
+
+  private String getAwsRegion() throws IOException {
+    // For AWS Lambda, the region is retrieved through the AWS_REGION environment variable.
+    String region = getEnv("AWS_REGION");
+    if (region != null) {
+      return region;
+    }
+
+    AwsCredentialSource awsCredentialSource = (AwsCredentialSource) this.credentialSource;
+    if (awsCredentialSource.regionUrl == null || awsCredentialSource.regionUrl.isEmpty()) {
+      throw new IOException(
+          "Unable to determine the AWS region. The credential source does not contain the region URL.");
+    }
+
+    region = retrieveResource(awsCredentialSource.regionUrl, "region");
+
+    // There is an extra appended character that must be removed. If `us-east-1b` is returned,
+    // we want `us-east-1`.
+    return region.substring(0, region.length() - 1);
+  }
+
+  @VisibleForTesting
+  AwsSecurityCredentials getAwsSecurityCredentials() throws IOException {
+    // Check environment variables for credentials first.
+    String accessKeyId = getEnv("AWS_ACCESS_KEY_ID");
+    String secretAccessKey = getEnv("AWS_SECRET_ACCESS_KEY");
+    String token = getEnv("Token");
+    if (accessKeyId != null && secretAccessKey != null) {
+      return new AwsSecurityCredentials(accessKeyId, secretAccessKey, token);
+    }
+
+    // Credentials not retrievable from environment variables - call metadata server.
+    AwsCredentialSource awsCredentialSource = (AwsCredentialSource) credentialSource;
+    // Retrieve the IAM role that is attached to the VM. This is required to retrieve the AWS
+    // security credentials.
+    if (awsCredentialSource.url == null || awsCredentialSource.url.isEmpty()) {
+      throw new IOException(
+          "Unable to determine the AWS IAM role name. The credential source does not contain the"
+              + " url field.");
+    }
+    String roleName = retrieveResource(awsCredentialSource.url, "IAM role");
+
+    // Retrieve the AWS security credentials by calling the endpoint specified by the credential
+    // source.
+    String awsCredentials =
+        retrieveResource(awsCredentialSource.url + "/" + roleName, "credentials");
+
+    JsonParser parser = OAuth2Utils.JSON_FACTORY.createJsonParser(awsCredentials);
+    GenericJson genericJson = parser.parseAndClose(GenericJson.class);
+
+    accessKeyId = (String) genericJson.get("AccessKeyId");
+    secretAccessKey = (String) genericJson.get("SecretAccessKey");
+    token = (String) genericJson.get("Token");
+
+    // These credentials last for a few hours - we may consider caching these in the
+    // future.
+    return new AwsSecurityCredentials(accessKeyId, secretAccessKey, token);
+  }
+
+  @VisibleForTesting
+  String getEnv(String name) {
```

Comment:
perhaps static

Reply:
Can't - needs to be overriden in unit tests